**Note to grader:** Each question consists of parts, e.g. Q1(i), Q1(ii), etc. Each part must be first graded  on a 0-4 scale, following the standard NJIT convention (A:4, B+: 3.5, B:3, C+: 2.5, C: 2, D:1, F:0). However, any given item may be worth 4 or 8 points; if an item is worth 8 points, you need to accordingly scale the 0-4 grade.


The total score must be re-scaled to 100. That should apply to all future assignments so that Canvas assigns the same weight on all assignments.



# **Assignment 1**



This assignment familiarizes you with Pytorch and reinforces the key steps in defining and training a simple regression model.

## Preparation Steps

In [1]:
# Import all necessary python packages
import numpy as np
import torch



## <font color = 'blue'> **Question 1. Basic Operations with Tensors** </font>

Your task for this question is to follow the NumPy  [**tutorial**](https://jalammar.github.io/visual-numpy/?fbclid=IwAR0tSntx5mj1aHteokRKrT4G6z77M3z0Quj40AQZ9mvKlhs2RTN3xXrc6Eo) (up to section **Data Representations**) and 'mirror' each of the operations presented in the tutorial with tensors in PyTorch.

You may find useful to consult this PyTorch introductory [tutorial](https://jhui.github.io/2018/02/09/PyTorch-Basic-operations/), and as always the full PyTorch [documentation](https://pytorch.org/docs/stable/torch.html) is the ultimate resource.

*(Please insert cells below for your answers )*

In [42]:
torch.manual_seed(33)

#### Initialization

In [43]:
x = torch.Tensor(3,3)
y = torch.Tensor([1,2,3])
z = torch.Tensor([[1,2,3],[4,5,6]])
a = torch.zeros(4,4)
b = torch.rand(4,4)
c = torch.ones(4,4)
r1 = torch.randn(4,4).type(torch.FloatTensor) # random numbers with SD = 1 and mean = 0
r2 = torch.randperm(4)
r3 = torch.ones([2,4], dtype=torch.float16)

print(x)
print(y)
print(z)
print(a)
print(b)
print(c)
print(r1)
print(r2)
print(r3)

tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
tensor([1., 2., 3.])
tensor([[1., 2., 3.],
        [4., 5., 6.]])
tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])
tensor([[0.6186, 0.5587, 0.1937, 0.3360],
        [0.2008, 0.6970, 0.6367, 0.1452],
        [0.8213, 0.2365, 0.3702, 0.9168],
        [0.0333, 0.7183, 0.0325, 0.2320]])
tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])
tensor([[-0.4415, -1.5009,  1.2530, -0.6902],
        [ 0.2395,  1.3365, -0.5728,  0.5368],
        [ 0.4898, -1.2608, -0.3192,  0.3207],
        [-0.1966, -0.7767, -1.4657, -1.0870]])
tensor([0, 3, 1, 2])
tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.]], dtype=torch.float16)


In [44]:
i = torch.eye(4) #identity matrix
v = torch.ones(2, 2, 2, 2)
v1 = torch.ones_like(i)
v2 = torch.arange(0,5,1)
print(i)
print(v)
print(v1)
print(v2)

tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
tensor([[[[1., 1.],
          [1., 1.]],

         [[1., 1.],
          [1., 1.]]],


        [[[1., 1.],
          [1., 1.]],

         [[1., 1.],
          [1., 1.]]]])
tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])
tensor([0, 1, 2, 3, 4])


#### Arithmetic (element-wise)

In [45]:
x.add_(y)
print(x)

tensor([[1., 2., 3.],
        [1., 2., 3.],
        [1., 2., 3.]])


In [46]:
d = torch.sub(x,y)
print(d)

tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])


In [47]:
e = torch.Tensor(3,3)
torch.mul(x,y,out=e)
print(e)

tensor([[1., 4., 9.],
        [1., 4., 9.],
        [1., 4., 9.]])


In [48]:
f = torch.div(x,y)
print(f)

tensor([[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]])


In [49]:
f.mul_(10)
print(f)

tensor([[10., 10., 10.],
        [10., 10., 10.],
        [10., 10., 10.]])


#### Indexing

In [50]:
print(e[:,0])
print(e[0,:])
print(e[0:3,1:3])
print(e[0,2])

tensor([1., 1., 1.])
tensor([1., 4., 9.])
tensor([[4., 9.],
        [4., 9.],
        [4., 9.]])
tensor(9.)


#### Split a tensor

In [51]:
r = torch.chunk(e,3)
print(r)

(tensor([[1., 4., 9.]]), tensor([[1., 4., 9.]]), tensor([[1., 4., 9.]]))


In [52]:
r = torch.split(r1,2)
print(r)

(tensor([[-0.4415, -1.5009,  1.2530, -0.6902],
        [ 0.2395,  1.3365, -0.5728,  0.5368]]), tensor([[ 0.4898, -1.2608, -0.3192,  0.3207],
        [-0.1966, -0.7767, -1.4657, -1.0870]]))


#### Reshape

In [53]:
r4 = r1.view(-1,2)
print(r4)

tensor([[-0.4415, -1.5009],
        [ 1.2530, -0.6902],
        [ 0.2395,  1.3365],
        [-0.5728,  0.5368],
        [ 0.4898, -1.2608],
        [-0.3192,  0.3207],
        [-0.1966, -0.7767],
        [-1.4657, -1.0870]])


In [54]:
E = e.reshape(-1,1)
print(E.shape)

torch.Size([9, 1])


In [55]:
X = torch.randn(3, 4)       
Y = X.unsqueeze(0)         
print(Y.shape)  

Z = Y.squeeze(0)
print(Z.shape)

torch.Size([1, 3, 4])
torch.Size([3, 4])


#### Transpose

In [56]:
r1_t = r1.t()
print(r1_t)

tensor([[-0.4415,  0.2395,  0.4898, -0.1966],
        [-1.5009,  1.3365, -1.2608, -0.7767],
        [ 1.2530, -0.5728, -0.3192, -1.4657],
        [-0.6902,  0.5368,  0.3207, -1.0870]])


In [57]:
r5 = torch.randn(2,3,4)
r5_t = r5.transpose(0,1)
print(r5_t.shape)

torch.Size([3, 2, 4])


#### Dot Product

In [58]:
r6 = torch.dot(y,y) #1d tensors only
print(r6)

tensor(14.)


#### Matrix - Vector Product

In [59]:
mat = torch.randn(3,4)
vec = torch.rand(4)
res = torch.mv(mat,vec)
print(res)

tensor([1.4434, 0.6357, 0.0170])


In [60]:
# Matrix + Mat*vec
M = torch.rand(3)
res = torch.addmv(M,mat,vec)
print(res)

tensor([1.6289, 1.0830, 0.5203])


#### Matrix - Matrix Product

In [61]:
m = torch.randn(3,3)
n = torch.randn(3,2)
res = torch.mm(m,n)
print(res)

tensor([[2.5321, 0.5235],
        [0.5326, 0.1301],
        [1.9316, 0.7628]])


In [62]:
# Matrix + Matrix*Matrix
M = torch.randn(3,2)
res = torch.addmm(M,m,n)
print(res)

tensor([[ 1.9501,  1.0562],
        [ 1.7942, -1.3221],
        [ 1.9689, -0.3867]])


#### Batch Matrix Multiplication

In [63]:
A = torch.randn(10, 3, 4)  # batch of 10
B = torch.randn(10, 4, 5)  
C = torch.bmm(A, B)        
print(C.shape)
print(C)

torch.Size([10, 3, 5])
tensor([[[ 1.8225,  1.4701,  1.9491,  1.0586, -1.0528],
         [ 0.4583,  1.3312,  1.6611,  0.4987,  1.3900],
         [-1.6847,  1.4925, -1.0693, -1.7511, -2.0955]],

        [[-0.3649, -0.9117,  3.0273,  2.9765, -2.0111],
         [ 0.7599,  2.6158,  1.3514, -1.3128,  0.8260],
         [ 0.3869,  0.6798,  3.4102,  1.9362, -2.1905]],

        [[-2.0007, -1.4244, -2.4081,  0.9214, -0.1149],
         [-1.0363, -0.2101,  0.7530, -3.8958, -1.8063],
         [-1.1981, -3.4992,  1.2785,  0.6312, -0.2503]],

        [[-4.2638, -1.6138,  2.0528, -1.8974, -1.2153],
         [ 6.4677, -1.9592, -0.8187,  4.9601,  4.9604],
         [-0.3094,  2.0407,  2.4481, -0.1072,  0.9820]],

        [[ 1.5655,  0.8277, -0.7178,  1.8265,  0.3514],
         [-0.1505,  1.6630,  0.6531,  1.3612,  0.9814],
         [ 4.8763,  0.0597, -3.8884,  2.5301, -1.7111]],

        [[ 2.4039,  0.2413, -0.3492,  0.5997,  1.3322],
         [-3.8226, -1.9117,  0.3516,  0.1737, -2.1312],
         [ 0.17

#### Statistics

In [64]:
largest = torch.max(r1)
print(largest)

tensor(1.3365)


In [65]:
smallest = torch.min(r1)
print(smallest)

tensor(-1.5009)


In [66]:
mean = torch.mean(r1)
print(mean)

tensor(-0.2584)


In [67]:
rowMeans = torch.mean(r1,dim=1)
print(rowMeans)
colMeans = torch.mean(r1,dim=0)
print(colMeans)

tensor([-0.3449,  0.3850, -0.1924, -0.8815])
tensor([ 0.0228, -0.5505, -0.2762, -0.2299])


#### Mean Squared Error

In [68]:
pred = torch.rand(10)
obs = torch.rand(10)
mse = torch.mean((pred-obs)**2)
print('MSE is ',mse)

MSE is  tensor(0.1495)


In [ ]:
# For grader use only

G = [0]*2


# insert grade here  (from 0 to 8)
# G[1] =

# please justify point subtraction  s

## <font color = 'blue'> **Question 2. Using the GPU** </font>

(i) Initialize randomly two 2D tensors, of dimensions 5000 x 5000 using torch.randn(). Multiply them 100 times on the CPU, and measure the total runtime. <br>
(ii) Activate and detect the GPU. Move the tensors to the GPU and repeat the above experiment. What do you observe? <br>
(iii) Continuing on the GPU, measure the runtime of a loop that does the following 1000 times:
*   (a) Define two random 100x100 matrices, A and B
*   (b) Multiply A and B
*   (c) Print the Frobenius norm of the product


(iv) Now do a very similar experiment. Measure the runtime of a loop that does the following 1000 times:
*   (a) Define two random 100x100 matrices, A and B
*   (b) Multiply A and B
*   (c) Store the Frobenius norm of the product in a 1D tensor $r$  that has been placed in the GPU. <br>
Finally, when exiting the loop, print the content of the tensor $r$. What do you observe, and how do you explain it?



In [80]:
t1 = torch.randn(5000,5000)
t2 = torch.randn(5000,5000)

In [82]:
import time

start = time.time()
for _ in range(100):
    res = torch.mm(t1,t2)
end = time.time()
time1 = end - start
print(f'Total time on CPU is {time1:.8f} seconds')

Total time on CPU is 9.72259831 seconds


In [2]:
if torch.backends.mps.is_available():
    device = torch.device('mps')
    print('Using MPS (Apple GPU)')

Using MPS (Apple GPU)


In [ ]:
t1 = t1.to(device)
t2 = t2.to(device)

torch.mps.synchronize() #ensures previous gpu work is done
start = time.time()
for _ in range(100):
    res = torch.mm(t1,t2)
torch.mps.synchronize() #waits for gpu work to finish
end = time.time()
time2 = end - start
print(f'Total time on {device} is {time2:.8f} seconds')

Total time on mps is 3.00121808 seconds


In [88]:
perfBoost = time1 / time2 
print(f'Performance boost : {perfBoost}x')

reductionPerc = (time1-time2)/time1 * 100
print(f'Reduction Percentage : {reductionPerc}%')

Performance boost : 3.2395507602028943x
Reduction Percentage : 69.13152242327052%


##### ii) Observations 
A performance boost, i.e, $cpuTime/mpsTime$ of <b>3.23x</b> with a relative reduction in computation time, i.e, $(cpuTime - mpsTime) / cpuTime$ of <b>69.13%</b>.

In [89]:
torch.mps.synchronize()
start = time.time()
for _ in range(1000):
    A = torch.rand(100,100,device=device)
    B = torch.rand(100,100,device=device)
    res = torch.mm(A,B)
    frobeniusNorm = torch.norm(res,p='fro')
    print(f"Frobenius norm: {frobeniusNorm}")
torch.mps.synchronize()    
end = time.time()
time3 = end - start
print(f'Total time on {device} is {time3:.8f} seconds')

Frobenius norm: 2482.5205078125
Frobenius norm: 2485.42041015625
Frobenius norm: 2528.246826171875
Frobenius norm: 2498.389404296875
Frobenius norm: 2518.50439453125
Frobenius norm: 2543.437744140625
Frobenius norm: 2498.574951171875
Frobenius norm: 2546.34033203125
Frobenius norm: 2517.173095703125
Frobenius norm: 2523.905029296875
Frobenius norm: 2490.16748046875
Frobenius norm: 2472.15380859375
Frobenius norm: 2516.241455078125
Frobenius norm: 2507.14404296875
Frobenius norm: 2512.6513671875
Frobenius norm: 2489.712646484375
Frobenius norm: 2496.268310546875
Frobenius norm: 2486.9794921875
Frobenius norm: 2511.893310546875
Frobenius norm: 2484.94677734375
Frobenius norm: 2491.867431640625
Frobenius norm: 2507.839111328125
Frobenius norm: 2540.99609375
Frobenius norm: 2552.773193359375
Frobenius norm: 2517.564697265625
Frobenius norm: 2505.604248046875
Frobenius norm: 2544.404052734375
Frobenius norm: 2507.220703125
Frobenius norm: 2496.630126953125
Frobenius norm: 2516.93603515625
F

In [90]:
r = torch.empty(1000,device=device)
torch.mps.synchronize()
start = time.time()
for i in range(1000):
    A = torch.rand(100,100,device=device)
    B = torch.rand(100,100,device=device)
    res = torch.mm(A,B)
    r[i] = torch.norm(res,p='fro')
torch.mps.synchronize()
end = time.time()
time4 = end - start
print(f'r is {r}')
print(f'Total time on {device} is {time4:.8f} seconds')

r is tensor([2524.5066, 2484.2039, 2498.2095, 2523.1694, 2515.9666, 2531.3979,
        2538.9333, 2507.5005, 2541.6570, 2504.9268, 2492.1843, 2516.7617,
        2499.8875, 2467.7434, 2533.5457, 2524.6824, 2502.3796, 2551.8398,
        2508.1826, 2547.2498, 2512.5317, 2537.7761, 2495.5359, 2503.7341,
        2478.9946, 2504.7781, 2533.9634, 2524.0483, 2520.4736, 2525.4404,
        2509.2285, 2567.3203, 2491.8943, 2507.2212, 2496.9939, 2507.1743,
        2509.4917, 2537.3740, 2540.9255, 2495.3743, 2511.5193, 2496.3228,
        2525.6792, 2482.6638, 2541.3545, 2503.2002, 2459.9624, 2541.5974,
        2537.7969, 2494.4751, 2530.2891, 2478.0254, 2475.0972, 2526.1047,
        2522.1677, 2502.6008, 2521.4973, 2449.2214, 2493.0222, 2506.3208,
        2491.9038, 2520.9895, 2509.9551, 2520.7104, 2498.9380, 2541.0068,
        2480.5369, 2521.6458, 2498.9905, 2466.4099, 2488.5676, 2515.8020,
        2487.0564, 2555.3262, 2500.5923, 2534.5068, 2496.8596, 2555.0125,
        2480.9990, 2514.2031, 253

In [91]:
perfBoost = time3 / time4 
print(f'Performance boost : {perfBoost}x')

reductionPerc = (time3-time4)/time3 * 100
print(f'Reduction Percentage : {reductionPerc}%')

Performance boost : 3.923782228828694x
Reduction Percentage : 74.514385822617%


#### iv) Observations
There's a performance boast of <b>3.92x</b> with a relative reduction percentage of <b>74.51%</b>. This is because print statements take time to execute and are done on the CPU. In the second situation where only one print statement is required in the end, we observe uninterrupted GPU utilization, while CPU-GPU synchronization was required in the former case, slowing down execution.

## <font color = 'blue'> **Question 3. Quadratic Regression** </font>

In the lecture we discussed a simple regression problem, where points are coming from the line $y = 2x + 1$, plus some noise. For this question you are asked to:

(i) Generate a dataset $D$ coming from a quadratic function: $y = -x^2 + 3x + 10$, plus some Gaussian noise with standard deviation 0.5. Here $x$ should range in $[-10,10]$ with 1000 data points. <br>

(ii) Create a dataloader for the set $D$, with a batch size of your choice. <br>

(iii) Define a model with a forward function that takes as input a scalar $x$, and returns a value $y = w_2 x^2 + w_1x + w_0$, where $w_i$ are the model parameters. The forward function should create polynomial features $[x, x^2]$ and apply a linear layer to produce the output. <br>

(iv) Write a training loop that adjusts the parameters of the model in order to fit the data $D$, using the Mean Squared Error loss and SGD optimizer, with a learning rate of 0.1. The training should take place on the GPU. <br>

(v) Train your model and report what parameter values it computes. (Sanity check: These should be close to $w_0 = 10$, $w_1 = 3$, $w_2 = -1$) <br>

(vi) Use a learning rate scheduler (e.g., StepLR or ExponentialLR) to decay the learning rate during training. <br>

(vii) The training loop should collect the values of the loss for each epoch. Visualize the loss by epoch of training for two experiments: with and without the scheduler. What do you observe? <br>

In [3]:
torch.manual_seed(33)

xTensor = torch.linspace(-10,10,1000)
epsilon = torch.randn(1000)*0.5


In [4]:
yTensor = -(xTensor)**2 + 3*xTensor + 10 + epsilon

In [5]:
xMean = xTensor.mean()
print('mean: ',xMean)
xStd = xTensor.std()
print('std: ',xStd)
normalizedX = (xTensor - xMean) / xStd
# did it because my values were exploding

mean:  tensor(-3.0518e-08)
std:  tensor(5.7822)


In [7]:
from torch.utils.data import Dataset, DataLoader
import torch.nn as nn
import torch.optim as optim

torch.manual_seed(33)

class CustomDataset(Dataset):
    def __init__(self, tensorX, tensorY):
        self.x = tensorX
        self.y = tensorY
    
    def __getitem__(self, index):
        return (self.x[index],self.y[index])
    
    def __len__(self):
        return len(self.x)
    
class QuadraticRegression(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(2,1)
        
    def forward(self,x):
        x = x.view(-1,1) 
        x2 = x**2
        features = torch.cat([x,x2],dim=1)
        return self.linear(features)    
    
def makeTrainStep(model,lossFunction,optimizer):
    def trainStep(x,y):
        model.train()
        pred = model(x)
        #print(pred.shape)
        y = y.view(-1,1)
        #print(y.shape)
        loss = lossFunction(pred,y)
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        return loss.item()
    return trainStep

def trainEpochs(epochs,trainLoader,trainStep,scheduler=None):
    losses = []
    for _ in range(epochs):
        epochLoss = 0.0
        for xBatch, yBatch in trainLoader:
            xBatch = xBatch.to(device)
            yBatch = yBatch.to(device)
            loss = trainStep(xBatch,yBatch)
            epochLoss += loss
        losses.append(epochLoss)
        if scheduler:
            scheduler.step()
    return losses


trainData = CustomDataset(normalizedX,yTensor)
trainLoader = DataLoader(trainData,batch_size=32,shuffle=True)

lossFunction = nn.MSELoss()

model = QuadraticRegression().to(device)
optimizer = optim.SGD(model.parameters(),lr=0.1)
print('Initial state without scheduler : ',model.state_dict())
trainStep = makeTrainStep(model,lossFunction,optimizer)
lossesWithoutScheduler = trainEpochs(10,trainLoader,trainStep)
print('Final state without scheduler : ',model.state_dict())

model2 = QuadraticRegression().to(device)
optimizer2 = optim.SGD(model2.parameters(),lr=0.1)
stepLR = optim.lr_scheduler.StepLR(optimizer2,step_size=1,gamma=0.1)
print('Initial state with StepLR scheduler : ',model2.state_dict())
trainStep = makeTrainStep(model2,lossFunction,optimizer2)
lossesWithStepLR = trainEpochs(10,trainLoader,trainStep,stepLR)
print('Final state with StepLR scheduler : ',model2.state_dict())

Initial state without scheduler :  OrderedDict({'linear.weight': tensor([[0.1677, 0.0831]], device='mps:0'), 'linear.bias': tensor([-0.4332], device='mps:0')})
Final state without scheduler :  OrderedDict({'linear.weight': tensor([[ 17.3803, -33.4197]], device='mps:0'), 'linear.bias': tensor([9.9856], device='mps:0')})
Initial state with StepLR scheduler :  OrderedDict({'linear.weight': tensor([[-0.5184, -0.0966]], device='mps:0'), 'linear.bias': tensor([-0.0819], device='mps:0')})
Final state with StepLR scheduler :  OrderedDict({'linear.weight': tensor([[ 17.3437, -32.0753]], device='mps:0'), 'linear.bias': tensor([7.9527], device='mps:0')})


Now, its important to note that the model learns an equation for xNormalized and y, not x and y. The equation learned by the model can be expressed as : <br>

$ y = xNorm * w1 + (xNorm^2) * w2 + bias$ where $xNorm = (x - \mu) / \sigma$

$ \implies y = ((x - \mu) / \sigma)* w1 + (((x - \mu) / \sigma)^2) * w2 + bias $

$ \implies y = x^2 * (w2 / \sigma^2) + x * (w1 / \sigma - 2*\mu*w2 / \sigma^2) + ( \mu^2*w2 / \sigma^2 - \mu * w1 / \sigma + bias ) $

where, <br>
weights [w1,w2]: tensor([[ 17.3803, -33.4197]], device='mps:0') <br>
bias: tensor([9.9856], device='mps:0') <br>
mean ($\mu$):  tensor(-3.0518e-08) <br>
std ($\sigma$):  tensor(5.7822) <br>

Given that the mean is so small, we approximate it to 0. 

$ \implies y = x^2 * (-33.4197 / (5.7822) ^ 2) + x * (17.3803 / 5.7822) + 9.9856 $

$ \implies y = x^2 * (-0.99) + x * 3.005 + 9.9856 $

In [19]:
import plotly.graph_objs as go
import plotly.offline as pyo

x = list(range(len(lossesWithoutScheduler)))  
trace1 = go.Scatter(x=x, y=lossesWithoutScheduler, mode='lines', name='Without Scheduler')
trace2 = go.Scatter(x=x, y=lossesWithStepLR, mode='lines', name='With StepLR Scheduler')

layout = go.Layout(title='Line Graph', xaxis=dict(title='Epochs'), yaxis=dict(title='Loss'))
fig = go.Figure(data=[trace1, trace2], layout=layout)

pyo.iplot(fig) 


#### Observation
The loss observed without using a scheduler is lower than the one with it possibly overshooting or converging harder.

In [ ]:
# for grader use only

# insert grade here (from 0 to 28)

# G[2] =
#
# please justify point subtractions

In [ ]:
# total score
max_score = 36
final_score = sum(G)*(100/max_score)